dartagnan/src/main/java/com/dat3m/dartagnan/encoding/PropertyEncoder.java

```diff
@@ -204,7 +204,22 @@ private BooleanFormula encodeLastCoConstraints() {
                     BooleanFormula sameAddress = context.sameAddress(init, w1);
                     Formula v2 = ExpressionEncoder.getLastMemValueExpr(init.getBase(), init.getOffset(), fmgr);
                     BooleanFormula sameValue = context.equal(context.value(w1), v2);
-                    enc.add(bmgr.implication(bmgr.and(lastCoExpr, sameAddress), sameValue));
+                    BooleanFormula other = bmgr.makeFalse();
+                    // Coherence is not guaranteed to be total in all models (e.g., PTX),
+                    // but the final value of a location should always match that of some coLast event.
+                    // No need to iterate over Inits since they are always ordered wrt coherence.
+                    // lastCo(w) => (lastVal(w.address) = w.val) 
+                    //           \/ (exists w2 : lastCo(w2) /\ lastVal(w.address) = w2.val))
+                    for (MemEvent w2 : program.getEvents(Store.class)) {
+                        if (!alias.mayAlias(w1, w2)) {
+                            continue;
+                        }
+                        BooleanFormula sameAddressOther = context.sameAddress(w1, w2);
+                        BooleanFormula lastCoExprOther = lastCoVar(w2);
+                        BooleanFormula sameValueOther = context.equal(context.value(w2), v2);
+                        other = bmgr.or(other, bmgr.and(lastCoExprOther, sameAddressOther, sameValueOther));
+                    }
+                    enc.add(bmgr.implication(bmgr.and(lastCoExpr, sameAddress), bmgr.or(sameValue, other)));
```

Comment:
Maybe we can make this encoding conditional on the fact whether `co` is total or not to avoid overhead `co` is total.
```
final boolean coIsTotal = (program.getArch() != Arch.PTX);
if (coIsTotal) {
  // usual encoding
} else {
  // new encoding
}
```
Apart from that, this encoding seems to be unnecessarily large. There are 3 nested loops but 2 should be sufficient (I think you create redundant constraints when `w1` and `w2` are swapped).
I think you can do roughly this:
```
for (Init init : program.getEvents(Init.class)) {
  BooleanFormula lastValueEnc = bmgr.mkFalse();
  BooleanFormula lastStoreExistsEnc = bmgr.mkFalse();
  for (Store w : program.getEvents(Store.class)) { // TODO: restrict to mayAlias only
      BooleanFormula isLast = ...;
      BooleanFormula sameAddr = ...;
      BooleanFormula sameVal = ...;
      lastValueEnc = bmgr.or(lastValueEnc, bmgr.and(isLast, sameAddr, sameVal));
      lastStoreExistsEnc = bmgr.or(bmgr.and(isLast, sameAddr));
  }
  enc.add(bmgr.ITE(lastStoreExists, lastValueEnc, context.equal(initValue, 0))); // Pseudo-code
}
```
This resembles your encoding, but leaves out the write of the outer loop (the code should therefore get moved outside of the outer loop). Also, I think you need to include `Init` events for otherwise if there is no single store to that address, the encoding will leave the final address value unconstrained rather than matching it with the init event.

Reply:
Updated

Reply:
Your changes look good, but you still miss the case where there exists no store to the same address as the init event. In that case, the initial value never gets overwritten so you should have `lastValue = initialValue`.
Your current encoding leaves the `lastValue` unconstrained, so it can take any arbitrary value.

Reply:
readFromInit added.